commit 5d749d0bbe811c10d9048cde6dfebc761713abfd
Author: Gwendal Grignou <gwendal@chromium.org>
Date:   Tue Mar 8 09:13:52 2016 -0800

    platform/chrome: cros_ec_dev - Fix security issue
    
    Prevent memory scribble by checking that ioctl buffer size parameters
    are sane.
    Without this check, on 32 bits system, if .insize = 0xffffffff - 20 and
    .outsize the amount to scribble, we would overflow, allocate a small
    amounts and be able to write outside of the malloc'ed area.
    Adding a hard limit allows argument checking of the ioctl. With the
    current EC, it is expected .insize and .outsize to be at around 512 bytes
    or less.
    
    Signed-off-by: Gwendal Grignou <gwendal@chromium.org>
    Signed-off-by: Olof Johansson <olof@lixom.net>

diff --git a/drivers/platform/chrome/cros_ec_dev.c b/drivers/platform/chrome/cros_ec_dev.c
index d45cd254ed1c..187470c8a1f6 100644
--- a/drivers/platform/chrome/cros_ec_dev.c
+++ b/drivers/platform/chrome/cros_ec_dev.c
@@ -137,6 +137,10 @@ static long ec_device_ioctl_xcmd(struct cros_ec_dev *ec, void __user *arg)
 	if (copy_from_user(&u_cmd, arg, sizeof(u_cmd)))
 		return -EFAULT;
 
+	if ((u_cmd.outsize > EC_MAX_MSG_BYTES) ||
+	    (u_cmd.insize > EC_MAX_MSG_BYTES))
+		return -EINVAL;
+
 	s_cmd = kmalloc(sizeof(*s_cmd) + max(u_cmd.outsize, u_cmd.insize),
 			GFP_KERNEL);
 	if (!s_cmd)